src/main.cpp

```diff
@@ -4488,7 +4484,16 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             {
                 CAddress addr = GetLocalAddress(&pfrom->addr);
                 if (addr.IsRoutable())
+                {
+                    LogPrintf("will push r %s", addr.ToString());
+                    pfrom->PushAddress(addr);
```

Comment:
If `AddLocal` in net.cpp was successfull e.g. by usage with `-upnp` then GetLocalAddress will always return this address. any code that calls AddLocal is only called during startup, thus LocalAddresses are not updated.
If only local addresses are discovered, e.g., without `-upnp` and behind NAT of the router, then no addresses are stored, and the check above will fail. Then the check of the following `else if` will pass, and an external IP is guessed from the other node.

Dynamic of IPs is not part of Bitcoin either.
The discovery Threads (UPNP and interface IP) has to be updated, that addresses are checked every 20 minutes or so, as UPNP uses this interval for portmapping anyway. Deletion of old IPs also needs to be added.